Added second systems engineering challenge. (#125)
gravitational/careers: challenges/systems/challenge-2.md

Split your code submission into roughly 3-5 Pull Requests to give the team an
opportunity to review your code and provide feedback. Feel free to merge each
PR after you have two approvals.

Our team will do their best to provide a high quality review of the submitted
Pull Requests in a reasonable time frame. You are spending your time on this,
we are going to contribute our time too.

After the final submission, the interview team will assemble and vote using a
"+1, -2" anonymous voting system: +1 is submitted whenever a team member
accepts the submission, -2 otherwise.

In case of a positive result, we will connect you to our HR and recruiting
teams, who will work out the details and present an offer.

In case of a negative score result, the hiring manager will contact you and
share a list of the key observations from the team that affected the result.

### Dependencies

Please write as much of your own code as possible. Avoid relying on
non-reputable dependencies for key components of the challenge.

# Requirements

## Level 4

Implement a CLI tool and backend (relying party) that can perform OIDC
Authorization Code Flow with PKCE against an OpenID Provider (also known as an

> identity provider or IdP) to obtain a valid JWT that can be used for

Reviewer: ```suggestion
identity provider or IdP) to obtain and print out a valid JWT that can be used for
```

    hash: "$2y$10$.Tx0T8URLvW.uMBPIZkJnuk7msXSTNWwJIajiOmuy4sLWysOtYOjW"

enablePasswordDB: true

oauth2:
  passwordConnector: local
```

Run Dex with the following Docker one-liner:

```
docker run --rm \
    -p 5557:5557 \
    -v $(pwd)/dex-config.yaml:/etc/dex/config.yaml \
    ghcr.io/dexidp/dex:v2.37.0 \
    dex serve --web-http-addr=0.0.0.0:5557 /etc/dex/config.yaml
```

Dex will now be available at http://idp.example.com:5557.

# Guidance

## Code and project ownership

This is a test challenge and we have no intent of using the code you've
submitted in production. This is your work, and you are free to do whatever you
feel is reasonable with it. In the scenario where you don't pass, you can open
source it with any license and use it as a portfolio project.

## Areas of focus

Teleport focuses on networking, infrastructure and security.

These are the areas we will be evaluating in the submission:

* Use consistent coding style. We follow
  [Go Coding Style](https://github.com/golang/go/wiki/CodeReviewComments) for
  the Go language. If you are going to use a different language, please pick

>   coding style guidelines and let us know what they are.

Author: Updated and moved `# Tools` up.

* Tests exist for happy path and error scenarios for key components of the
  challenge.
* Make sure builds are reproducible.
* Ensure error handling and error reporting is consistent. The system should
  report clear errors and not crash under non-critical conditions.
* Avoid concurrency and networking errors. Most of the issues we've seen in
  production are related to data races, networking error handling or goroutine
  leaks. We will be looking for those errors in your code.
* Security. Use strong authentication and a simple, but secure, authorization

>   scheme. Set up the strongest transport encryption you can. Test it.

Author: Updated wording.

# Summary

> Implement a prototype CLI tool that can perform OIDC authentication.

Reviewer: ```suggestion
Implement a prototype CLI tool that can perform OIDC Authorization Code Flow with PKCE.
```

Author: Reworded, can you review and tell me if you think the new wording makes more sense?

  we're doing here.

We believe this technique is not only better, but also is more fun compared to
whiteboard/quiz interviews so common in the industry. It's not without the
downsides - it could take longer than traditional interviews.

[Some of the best teams use coding challenges.](https://sockpuppet.org/blog/2015/03/06/the-hiring-post/)

We appreciate your time and are looking forward to hack on this project together.

# Levels

There are 6 engineering levels at Teleport. This challenge only applies to P4
applicants.

# Interview Process

The interview process will start with you receiving an invite to a private Slack
channel. That channel will contain the interview panel. You can ask them about
R&D culture, work-life balance, or anything else that you would like to learn
about Teleport.

## Design Doc

Before writing any actual code, we ask that you write a brief design document.
The design document should cover: design approach, scope, proposed API, security
considerations, CLI UX, and implementation details where appropriate. Start with
a brief doc that covers the edge cases and design approach. At Teleport, we
prefer Markdown for

> [our designs](https://github.com/gravitational/teleport/blob/master/rfd/0000-rfds.md).

Author: Done.

Once the design document has been approved by two reviewers, move on to the
implementation.

## Implementation

Split your code submission into roughly 3-5 Pull Requests to give the team an
opportunity to review your code and provide feedback. Feel free to merge each
PR after you have two approvals.

Our team will do their best to provide a high quality review of the submitted
Pull Requests in a reasonable time frame. You are spending your time on this,
we are going to contribute our time too.

After the final submission, the interview team will assemble and vote using a
"+1, -2" anonymous voting system: +1 is submitted whenever a team member
accepts the submission, -2 otherwise.

In case of a positive result, we will connect you to our HR and recruiting
teams, who will work out the details and present an offer.

In case of a negative score result, the hiring manager will contact you and
share a list of the key observations from the team that affected the result.

### Dependencies

Please write as much of your own code as possible. Avoid relying on non-first
party or reputable dependencies for key components of the challenge.

# Requirements

> ## Level 4

Author: Right now I am thinking this challenge is only valid for L4. If you want to apply at L5 you have to use the job worker challenge. After we get comfortable with this one, we can add additional scope to it to add an L5 section as well.

Implement a CLI tool that can perform OIDC authentication against an IdP to
obtain a valid JWT for authentication.

* Implement a CLI tool will be used to initiate OIDC login flow.
* Implement a backend service that will implement the OIDC callback handlers
  and perform authentication against an IdP.

> * Specifically implement the OIDC Authorization Code Flow with PKCE.

Reviewer: Nit: Feels like we should lead with this, I would consider rephrasing above to "Implement a CLI tool that can perform OIDC Authorization Code Flow with PKCE against an IdP ..." and remove this line.

* Use [Dex](https://dexidp.io) for your IdP using the below configuration and
  Docker one-liner with username/password: `foo`/`bar`.
* Obtain a valid JWT that _could_ be later used for authentication. For this

>   challenge simply printing the JWT to the screen in sufficient.

Author: Updated, can you take a look and tell me if this makes our expectations clear?

### Host and IdP configuration

Update DNS or `/etc/hosts` with the following entries:

```
127.0.0.1 backend.example.com
127.0.0.1 idp.example.com
```

Use Dex for your IdP. Save the configuration below as `dex-config.yaml`.

```

> issuer: "http://idp.example.com:5557"

Reviewer: Are we okay with using plain HTTP for the IdP for this challenge? The whole added security of PKCE flow relies solely on usage of TLS so the code verifier can't be intercepted so HTTP kinda defeats the purpose.

If we're ok with using HTTP for this, we should also probably call it out explicitly somewhere.

Author: We can ask people to use `mkcert` but not sure if it adds much to be honest.

Author: I mention in the tradeoffs section that it's okay to not use HTTPS just for this challenge and that on a real production system someone would use Let's Encrypt.

  bad idea for this task - try to stick to industry proven security methods as
  much as possible.

## Questions

It is OK to ask the interview team questions. Some folks stay away from asking
questions to avoid appearing less experienced, so we provide examples of
questions to ask and questions we expect candidates to figure out on their own.

Here is a great question to ask:

> Is it OK for the backend service to be served over HTTP? For a production
> system I would use Let's Encrypt to obtain TLS certificates and serve it over
> HTTPS. However, as a corner to cut I would like to serve it over HTTP.

It demonstrates that you thought about this problem domain, recognize the trade
off and are saving you and the team time by not implementing it.

This is the question we expect candidates to figure out on their own:

> What version of Go should I use? What dependency manager should I use?

Unless specified in the requirements, pick the solution that works best for you.

# Tools

This task should be implemented in Go and should work on 64-bit Linux machines.

# Timing

You can split coding over a couple of weekdays or weekends and find time to ask
questions and receive feedback.

> Once you join the Slack channel, you have between 1 to 2 weeks complete the
> challenge depending on the challenge you choose.

Reviewer: I don't think there's 1 week version?
```suggestion
Once you join the Slack channel, you have 2 weeks to complete the challenge.
```
